Fix reek offenses in SamlIdp-related files

monfresh opened the pull request:
**Why**: So our code smells better

sbc100 commented on .reek:
```diff
@@ -22,6 +21,7 @@ TooManyMethods:
 UncommunicativeMethodName:
   exclude:
     - Users::PhoneConfirmationController
+    - 'render_401'
```
Quotes needed for some yml reason/

sbc100 commented on .reek:
```diff
@@ -34,4 +34,4 @@ UtilityFunction:
 exclude_paths:
   - db/migrate
-  - spec/**/*
+  - spec
```
In the long run I'm not totally sure about this.  I think test code should smell good too.. maybe we should relax some of the rules only?

sbc100 commented on app/decorators/user_decorator.rb:
```diff
@@ -61,6 +61,46 @@ def no_longer_blocked_from_entering_2fa_code?
+    active_identities[1]
+  end
+
+  def last_identity
```
This stuff kinda feels like it *is* part of the data model to me, since its about the user->identity data relation, but maybe I'm wrong?  Convince me :)

sbc100 commented:
At first glace this PR looks much bigger than a "fixing smells" PR.   I know this might be a big ask, but do you think you could break out the user -> user_decorator refactor into separate PR (or at least a seperate CL, so that we can review that in isolation?).  I feel like that warrants a PR of its own with its own description.

monfresh commented:
Ready to review now!

sbc100 commented on .reek:
```diff
@@ -22,17 +26,26 @@ TooManyMethods:
     - AnalyticsEventJob#perform
     - SmsSenderConfirmationJob
+
+'spec':
```
What does this mean?   if it means what i think it means shouldn't `spec` be removed from the last line of this file?

monfresh replied:
This means apply the following rules to the `spec` path. Yes, the last line should be removed, but I wasn't sure if it would result in Code Climate reporting issues in the spec files, which would be unrelated to this PR. I should probably remove this change from this PR and do a separate one that configures Reek for the specs and fixes the spec offenses.

sbc100 commented on app/models/identity.rb:
```diff
@@ -2,7 +2,7 @@ class Identity < ActiveRecord::Base
 
-  def deactivate!
+  def deactivate
     update(last_authenticated_at: nil)
```
Wierdly I think this should be `update!` so we don't ignore failures... but deactivate doesn't need a `!`

pkarman replied:
yes please `update!`

sbc100 commented on app/models/user.rb:
```diff
@@ -47,11 +47,11 @@ def send_reset_confirmation
 
   def first_identity
-    active_identities[0] unless active_identities.empty?
+    active_identities[0]
```
Doesn't this need `|| NullIdentity.new` too?

monfresh replied:
Not according to the tests. `first_identity` is only called in the context of SLO, and a user will always have at least one active identity when SLO is initiated. The only exception is if the user is signed out and then attempts to log out. In that case, we can't call `user.first_identity` since there is no user. That is why we have `return NullIdentity.new unless user` in `SingleLogoutHandler#identity`.

sbc100 commented:
I would re-title this PR something like `Refactor SAML logout blah blah`.   I see this change as a lot more than fixing offenses.

sbc100 commented on app/controllers/concerns/saml_idp_logout_concern.rb:
```diff
@@ -1,129 +1,38 @@
-    # no SLO messages to generate; finish logout at IdP
-    return nil if user_session[:logout_response].nil?
+    LogoutResponseHandler.new(asserted_identity, user_session[:logout_response]).perform do |status|
+      return generate_slo_request if status == 'continue logout with next identity'
```
Why string comparisons here?  Why not constants like LogoutResponseHandler.CONTINUE_WITH_NEXT_ID?

monfresh replied:
The `LogoutResponseHandler` is yielding the state of the SLO flow to the controller so that the controller knows how to proceed. The status can be anything, and the controller needs to check the value of the status. The easiest thing to compare is either a string or a symbol. I thought a string was the most descriptive, but we could also do something like `if status == :continue_logout_with_next_identity`

pkarman replied:
I'd rather we used a symbol or a constant, preferably the latter. Makes typos a fatal error which in this case would be good.

sbc100 commented:
I'm not sure I completely understand this change, but it mostly looks good to me.

@pkarman should probably take a look too.

monfresh commented:
In order to fix Reek offenses, methods had to be extracted and/or moved to different classes, but I agree that the PR title can be renamed.

pkarman commented on app/services/single_logout_handler.rb:
```diff
@@ -0,0 +1,76 @@
+    @_identity ||= begin
+      first_identity = user.first_identity
+
+      return first_identity if
```
maybe Reek disagrees with me, but I find these `return foo if bar && baz` syntax constructs harder to maintain in the long run. I personally prefer
```
if saml_request && saml_request.issuer != first_identity.service_provider
  return
end
```

I know it's a bit longer but easier to drop a debug statement into later and easier to read, imo.

pkarman commented:
lgtm % comments.

monfresh commented:
@sbc100 @pkarman I addressed your feedback in [this commit](https://github.com/18F/identity-idp/pull/320/commits/a1b2f6a5eb9ca465928177e67539b710d2a0fddd). PTAL and let me know how it looks. I ended up using a symbol because a constant gave me warnings and requires defining the constant first and setting it to something, such as a string or a symbol, so might as well use the symbol directly.

pkarman commented:
Using a constant in a Struct will give warnings, yes. Why not make it a regular class? I don't see how a Struct is superior in this case.

```diff
diff --git a/app/controllers/concerns/saml_idp_logout_concern.rb b/app/controllers/concerns/saml_idp_logout_concern.rb
index c268145..93cfe72 100644
--- a/app/controllers/concerns/saml_idp_logout_concern.rb
+++ b/app/controllers/concerns/saml_idp_logout_concern.rb
@@ -12,9 +12,9 @@ module SamlIdpLogoutConcern
 
   def handle_saml_logout_response
     LogoutResponseHandler.new(asserted_identity, user_session[:logout_response]).perform do |status|
-      return generate_slo_request if status == :continue_logout_with_next_identity
+      return generate_slo_request if status == LogoutResponseHandler::CONTINUE_LOGOUT_RESPONSES
 
-      return finish_slo_at_idp if status == :no_more_logout_responses
+      return finish_slo_at_idp if status == LogoutResponseHandler::NO_MORE_LOGOUT_RESPONSES
     end
 
     generate_slo_response_and_sign_out
diff --git a/app/services/logout_response_handler.rb b/app/services/logout_response_handler.rb
index d268af4..b009979 100644
--- a/app/services/logout_response_handler.rb
+++ b/app/services/logout_response_handler.rb
@@ -1,12 +1,22 @@
-LogoutResponseHandler = Struct.new(:identity, :response) do
+class LogoutResponseHandler
+  NO_MORE_LOGOUT_RESPONSES = 1
+  CONTINUE_LOGOUT_RESPONSES = 2
+
+  attr_accessor :identity, :response
+
+  def initialize(identity, response)
+    @identity = identity
+    @response = response
+  end
+
   def perform
     deactivate_identity
 
-    yield :continue_logout_with_next_identity if in_slo?
+    yield CONTINUE_LOGOUT_RESPONSES if in_slo?
 
     deactivate_last_identity
 
-    yield :no_more_logout_responses if response.blank?
+    yield NO_MORE_LOGOUT_RESPONSES if response.blank?
   end
 
   private
```

monfresh commented:
I don't see the benefit of using a constant. It's more code and more assignments, and I think `LogoutResponseHandler::CONTINUE_LOGOUT_RESPONSES` is harder to read. Convince me?

sbc100 commented:
I think the main argument is it avoids typos.    But symbols are at least better than strings.

monfresh commented:
I don't understand the typos argument. If you misspell the symbol or string, the tests will fail either way.

pkarman commented:
It feels similar to using named exceptions rather than comparing error strings: you'll get a run-time error if you do this:
```
return finish_slo_at_idp if status == NO_MORE_LOOGUT_RESPONSES
```

but if you do this, you'll get no error, just unexpected behavior:
```
return finish_slo_at_idp if status == :no_more_loogut_responses
```

Depending on the tests is good. If I make a typo in my test when refactoring later, it might take me an hour to see it.

monfresh commented:
I'm going to refactor this to avoid the yield altogether. Commit coming up.

monfresh commented:
Here you go: https://github.com/18F/identity-idp/pull/320/commits/3ad8a69254f6d2617ad970e5fa3b7a7cc0f195cd

sbc100 commented:
@monfresh you are is true believer in dynamc languages :)   Who needs compile time errors when you have tests.   Of course you are correct and we should have full test coverage, but....

sbc100 commented on app/services/single_logout_handler.rb:
```diff
@@ -0,0 +1,76 @@
+      end
+      # Logout was initiated out of chronological order. Resume SLO
+      # with next Identity
+      return second_identity if second_identity
```
What do you think of `return second_identity || user.last_identity`?  (does it even work i'm not 100% sure)

monfresh replied:
Good catch. That will probably work. I'll give it a try and push a new PR.

pkarman commented:
I like this latest approach much better. Easier to grok.

I wonder if we can't clean up the handler even further to avoid some of the roundabout logic, but that can be in a separate PR. I'm fine to land this as-is.

For future, the SLO process logic is basically a slow-motion async while() loop, where each active identity is getting popped off the stack and logged out, until we're all done.

```ruby
if user.active_identities.any?
  do_logout
else
  finish_logout
end
```

seems like we could refactor the code to reflect that more clearly.

sbc100 commented:
Looks good in general.  I'll let peter give final lg this as he understand the logic better than me.

monfresh commented:
@pkarman Yeah, I agree the logic seems complicated. I tried to slim it down but my first attempt didn't work and didn't want to spend too much more time on it to get this PR out.
